modin/core/storage_formats/pandas/query_compiler.py

```diff
@@ -3026,6 +3060,99 @@ def groupby_dtypes(
             drop=drop,
         )
 
+    @_inherit_docstrings(BaseQueryCompiler.groupby_agg)
+    def _groupby_shuffle(
+        self,
+        by,
+        agg_func,
+        axis,
+        groupby_kwargs,
+        agg_args,
+        agg_kwargs,
+        drop=False,
+        how="axis_wise",
+    ):
+        if Engine.get() == "Python":
+            raise NotImplementedError(
+                "Reshuffling groupby is not implemented for python engine, see: "
+                + "https://github.com/modin-project/modin/issues/5916"
+            )
+
+        # Defaulting to pandas in case of an empty frame as we can't process it properly.
+        # Higher API level won't pass empty data here unless the frame has delayed
+        # computations. FIXME: We apparently lose some laziness here (due to index access)
+        # because of the inability to process empty groupby natively.
+        if len(self.columns) == 0 or len(self.index) == 0:
+            return super().groupby_agg(
+                by, agg_func, axis, groupby_kwargs, agg_args, agg_kwargs, how, drop
+            )
+
+        if isinstance(by, type(self)) and drop:
```

Comment:
Why does the conversion depend on `drop` parameter?

Reply:
well, it's legacy of the bad naming for the `drop` parameter. It actually means whether the `by` data came from the same frame:
https://github.com/modin-project/modin/blob/ab90115e5493b36a173539cfa2bbc68086dbfb52/modin/core/storage_formats/base/query_compiler.py#L2788-L2790

If `drop == True` then we know that `by` describes columns of the `self` and so we can replace it with the column names themselves.